workbench/src/workbench.c

```diff
@@ -0,0 +1,804 @@
+{
+	if (wb != NULL)
+	{
+		guint index;
+		gchar *current;
+
+		for (index = 0 ; index < wb->bookmarks->len ; index++)
+		{
+			current = g_ptr_array_index(wb->bookmarks, index);
+			if (current == filename)
+			{
+				g_ptr_array_remove_index (wb->bookmarks, index);
+				wb->modified = TRUE;
+				return TRUE;
+			}
+		}
+	}
+	return FALSE;
+}
+
+
+/** Get the bookmark at @a index.
+ *
+ * @param wb    The workbench
+ * @param index The index
+ * @return Address of filename or NULL
+ *
+ **/
+gchar *workbench_get_bookmark_at_index (WORKBENCH *wb, guint index)
+{
+	if (wb != NULL)
+	{
+		gchar *file;
+		file = g_ptr_array_index(wb->bookmarks, index);
+		if (file == NULL)
+		{
+			return NULL;
+		}
+		return file;
+	}
+	return NULL;
+}
+
+
+/** Get the number of bookmarks in a workbench.
+ *
+ * @param wb The workbench
+ * @return The number of bookmarks or 0 if wb == NULL
+ *
+ **/
+guint workbench_get_bookmarks_count(WORKBENCH *wb)
+{
+	if (wb != NULL && wb->bookmarks != NULL)
+	{
+		return wb->bookmarks->len;
+	}
+	return 0;
+}
+
+
+/** Save a workbench.
+ *
+ * @param wb    The workbench
+ * @param error Location for returning an GError
+ * @return TRUE on success, FALSE otherwise
+ *
+ **/
+gboolean workbench_save(WORKBENCH *wb, GError **error)
+{
+	gboolean success = FALSE;
+
+	if (wb != NULL)
+	{
+		GKeyFile *kf;
+		guint	 index;
+		gchar    *contents;
+		gchar    group[20];
+		gsize    length, boomarks_size;
+		WB_PROJECT_ENTRY *entry;
+
+		kf = g_key_file_new ();
+
+		/* Save common, simple values */
+		g_key_file_set_string(kf, "General", "filetype", "workbench");
+		g_key_file_set_string(kf, "General", "version", "1.0");
+		g_key_file_set_boolean(kf, "General", "RescanProjectsOnOpen", wb->rescan_projects_on_open);
+
+		/* Save Workbench bookmarks as string list */
+		boomarks_size = workbench_get_bookmarks_count(wb);
+		if (boomarks_size > 0)
+		{
+			gchar **bookmarks_strings, *file, *rel_path;
+
+			bookmarks_strings = g_new0(gchar *, boomarks_size+1);
+			for (index = 0 ; index < boomarks_size ; index++ )
+			{
+				file = workbench_get_bookmark_at_index(wb, index);
+				rel_path = get_any_relative_path(wb->filename, file);
+
+				bookmarks_strings[index] = rel_path;
+			}
+			g_key_file_set_string_list
+				(kf, "General", "Bookmarks", (const gchar **)bookmarks_strings, boomarks_size);
+			for (index = 0 ; index < boomarks_size ; index++ )
+			{
+				g_free (bookmarks_strings[index]);
+			}
+			g_free(bookmarks_strings);
+		}
+
+		/* Save projects data */
+		for (index = 0 ; index < wb->projects->len ; index++)
+		{
+			entry = g_ptr_array_index(wb->projects, index);
+			g_snprintf(group, sizeof(group), "Project-%u", (index+1));
+			g_key_file_set_string(kf, group, "AbsFilename", entry->abs_filename);
+			g_key_file_set_string(kf, group, "RelFilename", entry->rel_filename);
+			g_key_file_set_boolean(kf, group, "UseAbsFilename", entry->use_abs);
+		}
+		contents = g_key_file_to_data (kf, &length, error);
```

Comment:
in the (unlikely) event `g_key_file_to_data()` fails, you should make sure to properly handle it, because it is invalid to set a `GError` on top of another, and you reuse `error` in the `g_file_set_contents()` call below.